### ISSUE TITLE
Monster upgrades rework plan

### ISSUE THREAD

**Issue by drbig** [CleverRaven/Cataclysm-DDA#12805]
So I was planning to start with how things are now *yadda yadda*. I'll just write down my idea of how it should look like logic- and implementation-wise. Didn't want to pollute #12313 though that issue is obviously relevant.

 - Upgradable monsters get a boolean `upgrades` property. This is the one-and-only property that clearly defines if a monster can upgrade or not.
 - If a monster has `upgrades = true` then it also has to (and this will be checked on load) have exactly:
  - Either `upgrades_into` or `upgrades_group` (same functionality as now)
  - Either `upgrades_half_life` or `upgrades_chance` (same functionality as now)
  - `upgrades_period` - which specifies a time-span in days
 - Monsters can upgrade only *once* per `upgrades_period`, missed windows don't stack up. The only delay to monster upgrades is the starting season offset which guarantees that upgrade windows always start at the same time the player started. Also this means additional property on monsters to keep track of whether it has mutated already in the current period. (periods are continuous, e.g.: with `upgrade_period` of 14 days, Z upgraded on day 7, the upgraded Z didn't upgrade between 14 - 28 days, but it did on day 32, next windows is 42 - 56).
 - Monsters have a chance to stop upgrading once and for all (`upgrades = false` then). My current idea here is that if a monster passed upgrade chance check it will have 70/30 chance to either actually upgrade or stop upgrading.
 - There are only two ways to trigger monster update: the special attack and on load
   - The special attack obeys once per period rule, but can't be turned off and the upgrade_period is not tweakable by the player
   - The on load upgrade can be turned off and `upgrades_period` is tweakable by the player. The `MONSTER_UPGRADE_SCALE` works like this: values 0.01 - 0.99 make monsters upgrade (much) more often; values 1.01 - 100.00 make monsters upgrade (much) less often; 0.00 turns off on load upgrades completely; 1.00 means the default `upgrade_period`. This means we can call it "Monster upgrade scale" and actually have it work like that, e.g. 2.00 means on average it will take monsters twice as much time to fully upgrade (it should also make it simple for the player to scale this season length-wise; 2x longer season = 2.00 upgrade delay).

The above is obviously open for discussion and tweaking. The main goal is to have the monster upgrades smooth and sane code- and logic-wise.

**Comment by drbig** [comment]
Continuous upgrade window clarification:

1. Z with `upgrade_period = 7` upgrades on day 4 into a Z with `upgrade_period = 14` - the now-upgraded Z's 14 days periods start at day 7.

**Comment by vache** [comment]
Why not have the json be something like
```
"name" : "z",
"upgrades" : { 
   "into" : "z_upgrade", // OR "into_group" : "z_upgrade_group"
   "halflife" : 50, // OR "chance" : 25
   "period" : 3
},
"drops" : "drop_group",
etc...
```
that way all the upgrade info is contained in one place, if the "upgrades" member exists, then check it for the rest, if not, then keep going.

**Comment by drbig** [comment]
@vache I do agree.

**Comment by kevingranade** [comment]
Is there a reason to support both half-life and chance?
The 'only once per period and missed periods are ignored' concept, means
that monsters in areas you move through frequently will upgrade faster than
monsters in areas you've never been.  This makes it where migrating will
always reset the monster progression to period 2.
Simply maintaining a rolling half-life will do what we want, when a monster
is created or is upgraded, calculate when it will upgrade based on its
half-life and random input, stash the result, it doesn't need to be
recalculated unless a special ability changes it.  On load, check that
timestamp, if it is in the past, upgrade (this implies recalculating the
next upgrade time, and potentially applying it immediately).  Not sure
offhand the best way to make that interact with the master ability, but the
standard progression is more important.

**Comment by Coolthulhu** [comment]
Sounds good to me. Simple and probably easy to maintain.

Master's ability could be easily rewritten to support on-generation evolution timer simply by making it decrease said number if it's within an acceptable range.

**Comment by kevingranade** [comment]
Tiny tweak, all you need to keep track of in the monster itself is that timestamp, to disable updating you can just set it to MAX_INT.

**Comment by drbig** [comment]
@kevingranade:

> Is there a reason to support both half-life and chance?

None that I see except for "there is code for it already".

> monsters in areas you move through frequently will upgrade faster than
monsters in areas you've never been

Indeed. I was probably way too focused on 'everything upgrades on sight'. Note that if I stack missed windows all monsters will on average upgrade at the same rate.

> Simply maintaining a rolling half-life will do what we want

Interesting approach, but I'm not sure if I understand it. This is how I get it: on monster create calculate the upgrade_daystamp as rng(0, upgrade_period); on monster load check if current_day > upgrade_daystamp, if so: monster rolls the upgrade; if it upgraded calculate the upgrade_daystamp for the new monster as old_monster->upgrade_period + rng(0, upgrade_period), check if it is in the past, if so roll upgrade again; and so on until it's "no upgrade this check" or rolled "never upgrade again". Does that sound like what you have on mind?

**Comment by narc0tiq** [comment]
> if it upgraded calculate the upgrade_daystamp for the new monster as old_monster->upgrade_period + [...]

The problem with that is (I imagine) you wouldn't have a reference to the old monster if you weren't doing the roll right as you did the upgrade. You should probably pretend the old monster stops existing and calculate based on what the new monster is.

Otherwise, it matches my reading of Kevin's statements, and sounds like a reasonable setup.

**Comment by i2amroy** [comment]
Pretty sure a rolling half life is what we already do (or at least it's what I tried to make it do). Game calculates how many half lives have elapsed since the last time it rolled for an upgrade, then make that roll based on that.

**Comment by drbig** [comment]
@i2amroy upgrade roll-wise it does indeed. The problem is with update_check running too often, e.g. daily if you move around enough. Point here is to have the upgrade checks spread in time in a controllable manner. Among other things.

**Comment by i2amroy** [comment]
Fair enough.

**Comment by kevingranade** [comment]
Half-life means the time at which you expect half of a population to have
decayed, what I'm suggesting is that you flip that around and make a roll
for *when* the monster you are considering is going to "decay" when the
monster is first created, I'm sure there are examples for the formula to
derive this scattered around on the net.
There is no 'period', each monster has a preordained time they will upgrade
set at creation time, but we just don't tell the player when that will be.
The benefits of this are a minimal number of rolls, minimal moving parts,
and good precision. Once you upgrade, you know exactly when the upgrade
threshold was passed, so you can use it to determine exactly when the next
upgrade should be.
You can still arbitrarily adjust the half-life of a monster with an option
to accelerate or slow down upgrades, and a master can manipulate the time
to upgrade.

**Comment by drbig** [comment]
Ok, so taking all the above in:

The JSON format for upgrades becomes:
```
"upgrades" : { 
   "into" : "z_upgrade", // OR "into_group" : "z_upgrade_group"
   "halflife" : 50 // in days
},
```

The `MONSTER_UPGRADE_FACTOR` stays as stated (it multiplies half-life).

Now we do treat the half-life exactly like half-life: at start_time+half-life we expect half of the monsters to be already upgraded, hence each monster has exactly 50% chance of upgrade within the half-life period.

So on first load we can do: loop one_in(2) and keep accumulating the half-life for failed throws; on successful throw calc the upgrade_time as start_time+accumulated half-life+rng(0, half-life); on upgrade_check() see if we're past the upgrade_time; if yes upgrade monster and do the above loop for it to get its next upgrade_time (only if that upgraded form can still upgrade) - and if it's before current day repeat the whole process.

We handle the never_upgrade case by putting a limit on the loop e.g. max 5 iterations (which gives us 3.125% chance of monster getting going the never_upgrade path).

Does the above make sense?

As for the master upgrade: maybe they should halve the upgrade time? or have 50/50 chance of either halving the upgrade_time or making the monster upgrade immediately?

**Comment by kevingranade** [comment]
That will work, but I'm sure there's a simple formula where we can
calculate when the creature will upgrade down to the turn, otherwise we'll
have weird generations where half the zombies all upgrade at the same time,
then half the rest will all upgrade at the same time, and so on.

**Comment by drbig** [comment]
Maybe I'm missing something? rng(0, half-life) should spread the updates, and then the monsters can actually only upgrade on load, which means technically they will all upgrade at the same time anyway (e.g. exploring some new area after a year from start pretty much guarantees most Zs will undergo a number of upgrades straight on load, which is what we want).

**Comment by narc0tiq** [comment]
I think it's just that algorithms are hard to both explain and read in English. I won't and can't say your proposed algorithm will avoid doing weird things, and I think @kevingranade is trying to express the same thing. It sounds complicated, but it might be just fine.

**Comment by drbig** [comment]
Having #12828 effectively in: do we keep the uniform `rng()` distribution, do we switch to `std::exponential_distribution`, or something else?

**Comment by i2amroy** [comment]
Keeping it just the basic rng seems like it might work best. An exponential seems like you might end up with "chunky" upgrades, where it seems the vast majority of zombies upgrade every 14 days or so, instead of a smoother distribution.

As for failure to upgrade chance, why not just put a ```"stoppage_percent":X``` into the JSON? Then you just have to add two bools (which in a bitfield would be just 2 bits) to each monster. The first checks if a monster has already done it's failure chance, and the second one says if it passed or not. On loading a monster for the first time you roll a single percentage roll and if they fail then they don't upgrade anymore. If they pass they keep trying to upgrade until they eventually pass, which resets them so the next time they load they have to check for upgrade failure again. Simple, and more configurable percentage wise then a "only gets 5 upgrade roll chances" or something similar.

**Comment by drbig** [comment]
@i2amroy the current code has the `#define UPGRADE_MAX_ITERS 5` commented which does the never_upgrade scaled to half_life.

I'm certainly open changes, but I had hard time trying to visualise exp dist impact on the upgrades. Even though it makes more sense pure-physics wise (but we have a game here).

**Comment by kevingranade** [comment]
The exponential function would be less "chunky" because at the same time it'd be totally reasonable to change the resolution to # of turns, as it is the resolution is per day, which is pretty far from smooth.
Config-wise at least, you can't beat it for simplicity, "after this many turns, half of the monsters of this type will have upgraded."
Note that while I'm much in favor of this approach, the existing approach isn't "broken" per se, so I'm not particularly insistent about it.

Side issue: "into" could also be a list of monster/weight pairs, so at upgrade a monster could go down one of several paths, i.e. a graph of upgrades.

**Comment by drbig** [comment]
> Side issue: "into" could also be a list of monster/weight pairs, so at upgrade a monster could go down one of several paths, i.e. a graph of upgrades.

First: longer upgrade chains. We can add branching as an expansion to this mechanic - and I'll be all for it.

I'd actually tone down special Zs spawns a lot in favour of longer upgrade chains. After all aren't special Zs supposed to be... special? :)

**Comment by kevingranade** [comment]
Precisely, once this is working as intended, my plan is to only have certain base zombie types spawn at game start.

**Comment by i2amroy** [comment]
>@i2amroy the current code has the #define UPGRADE_MAX_ITERS 5 commented which does the never_upgrade scaled to half_life.

Yeah, I was just pointing out a future expansion idea that allows for the "stops upgrading at a given level" feature with a bit less hardwiring things into the game.

**Comment by kevingranade** [comment]
If we have a list of "upgrades into", one of the entries could be "stop upgrading", then you can finely tune the proportion of a population that never upgrades.

**Comment by i2amroy** [comment]
That would be a functionally equivalent way of doing it, yeah.

**Comment by Headjack** [comment]
Just to get this in at an early stage, is upgrading planned to be togglable vs. the traditional method?  I do think it's a cool concept, but the sheer number of hulks in my last life left me with a sour taste.

**Comment by narc0tiq** [comment]
It should already be disableable -- in the world settings, there's an
option for "Monster half-life scaling factor" that notes "Set to 0.00 to
turn off monster upgrades".

**Comment by Headjack** [comment]
Ok, just checking. i've been out of cata a while but the last exp I played was just starting to use it. Thx.

**Comment by kevingranade** [comment]
Keep in mind once it's done turning it off is going to be tantamount to turning on "vanilla zombies mode".

**Comment by Coolthulhu** [comment]
We could have the evolution as two settings: timer scaling and offset. That way only a percentage of population would be evolved.

**Comment by kevingranade** [comment]
Can you expand on that? I'm not sure what you mean.

**Comment by Barhandar** [comment]
Zombie upgrading would be a touch less crazy if regular zombies used a different list from what Zombie Masters use. If "freq" means what I think it does, that list is _heavily_ biased in favor of higher upgrade levels.

Replacing that with a list which only has basic types - hollows, skeletons, grabbers, spitters, toughs, shriekers, boomers, and hunters - would already reduce hordes of hulks and predators running amok.

**Comment by illi-kun** [comment]
We have monsters evolution implemented. Do we need to keep this open?
